File: strategy/typescript/Strategy.ts
// Strategy interface
interface PaymentStrategy {
  pay(amount: number): string;
}

// Concrete Strategies
class CreditCardPayment implements PaymentStrategy {
  constructor(private cardNumber: string) {}

  pay(amount: number): string {
    const lastFour = this.cardNumber.slice(-4);
    return `Paid $${amount.toFixed(2)} using Credit Card ending in ${lastFour}`;
  }
}

class PayPalPayment implements PaymentStrategy {
  constructor(private email: string) {}

  pay(amount: number): string {
    return `Paid $${amount.toFixed(2)} using PayPal account ${this.email}`;
  }
}

class BitcoinPayment implements PaymentStrategy {
  constructor(private walletAddress: string) {}

  pay(amount: number): string {
    const shortAddress = this.walletAddress.substring(0, 8) + '...';
    return `Paid $${amount.toFixed(2)} using Bitcoin wallet ${shortAddress}`;
  }
}

// Context
class ShoppingCart {
  private paymentStrategy: PaymentStrategy;

  constructor(private amount: number) {}

  setPaymentStrategy(strategy: PaymentStrategy): void {
    this.paymentStrategy = strategy;
  }

  checkout(): string {
    return this.paymentStrategy.pay(this.amount);
  }
}

// Usage
const cart = new ShoppingCart(100.0);

cart.setPaymentStrategy(new CreditCardPayment('1234567890123456'));
console.log(cart.checkout());

cart.setPaymentStrategy(new PayPalPayment('user@example.com'));
console.log(cart.checkout());

cart.setPaymentStrategy(new BitcoinPayment('1A2B3C4D5E6F7G8H'));
console.log(cart.checkout());

File: strategy/java/Strategy.java
// Strategy interface
interface PaymentStrategy {
    String pay(double amount);
}

// Concrete Strategies
class CreditCardPayment implements PaymentStrategy {
    private String cardNumber;

    public CreditCardPayment(String cardNumber) {
        this.cardNumber = cardNumber;
    }

    @Override
    public String pay(double amount) {
        String lastFour = cardNumber.substring(cardNumber.length() - 4);
        return String.format("Paid $%.2f using Credit Card ending in %s", amount, lastFour);
    }
}

class PayPalPayment implements PaymentStrategy {
    private String email;

    public PayPalPayment(String email) {
        this.email = email;
    }

    @Override
    public String pay(double amount) {
        return String.format("Paid $%.2f using PayPal account %s", amount, email);
    }
}

class BitcoinPayment implements PaymentStrategy {
    private String walletAddress;

    public BitcoinPayment(String walletAddress) {
        this.walletAddress = walletAddress;
    }

    @Override
    public String pay(double amount) {
        String shortAddress = walletAddress.substring(0, 8) + "...";
        return String.format("Paid $%.2f using Bitcoin wallet %s", amount, shortAddress);
    }
}

// Context
class ShoppingCart {
    private double amount;
    private PaymentStrategy paymentStrategy;

    public ShoppingCart(double amount) {
        this.amount = amount;
    }

    public void setPaymentStrategy(PaymentStrategy strategy) {
        this.paymentStrategy = strategy;
    }

    public String checkout() {
        return paymentStrategy.pay(amount);
    }
}

// Main class demonstrating Strategy pattern
public class Strategy {
    public static void main(String[] args) {
        ShoppingCart cart = new ShoppingCart(100.0);

        cart.setPaymentStrategy(new CreditCardPayment("1234567890123456"));
        System.out.println(cart.checkout());

        cart.setPaymentStrategy(new PayPalPayment("user@example.com"));
        System.out.println(cart.checkout());

        cart.setPaymentStrategy(new BitcoinPayment("1A2B3C4D5E6F7G8H"));
        System.out.println(cart.checkout());
    }
}
